src/org/zaproxy/zap/extension/csphelper/CspHelper.java

```diff
@@ -0,0 +1,467 @@
+		for (Element tag : tags) {
+			Attributes tagAttr = tag.getAttributes();
+			if (tagAttr != null) {
+				for (String attr : this.jsEventAttrs) {
+					Attribute jsAttr = tagAttr.get(attr);
+					if (jsAttr != null) {
+						csp.getScriptSource().addHashNonce("'" + 
+							ExtensionCspHelper.sha256(jsAttr.getValue()) +
+							"'");
+						logger.debug("SBSB adding JS attr hash " + ExtensionCspHelper.sha256(jsAttr.getValue()) + " (" + jsAttr.getValue() + ")");
+					}
+				}
+			}
+		}
+	}
+	
+	private void handleScripts(CSP csp, HttpMessage msg, Source source) {
+		// Start with the script tags
+		List<Element> scripts = source.getAllElements("script");
+		logger.debug("SBSB got scripts " + scripts.size()); // TODO
+		for (Element script : scripts) {
+			String src = script.getAttributeValue("src");
+			if (src != null) {
+				logger.debug("SBSB adding url " + src); // TODO 
+				csp.getScriptSource().addUrl(src);
+			} else {
+				// Handle inline scripts
+				String content = script.getContent().toString();
+				csp.getScriptSource().addHashNonce("'" + 
+					ExtensionCspHelper.sha256(content) +
+					"'");
+				logger.debug("SBSB adding hash " + ExtensionCspHelper.sha256(script.getContent().toString())); // TODO
+				handleWorkerSources(content, csp);
+				handleConnectSources(content, csp);
+			}
+		}
+		logger.debug("SBSB csp now " + csp.getSite() + " : " + csp.generate()); // TODO
+	}
+	
+	private void handleWorkerSources(String content, CSP csp) {
+		// Regex checks to identify connect-src, worker-src etc
+		ArrayList<String> workers = regexGetGroup("(Worker|\\.register)\\([\"']([\\w\\.:/?=&]*)[\"']", 2, content);
+		for (String w : workers) {
+			csp.getWorkerSource().addUrl(w);
+		}
+	}
+	
+	private void handleConnectSources(String content, CSP csp) {
+		// Regex checks to identify connect-src
+		// (WebSocket|EventSource|\\.sendBeacon)\\([\"']([\\w\\.:/?=&#]*)[\"']
+		// \\.open\\([\"'][a-zA-Z]+[\"'],\\s*[\"']([\\w\\.:/?=&#]*)[\"']
+		ArrayList<String> connects = regexGetGroup("[\"']((http[s]?:\\/\\/)?\\w+\\.\\w[\\.\\w/?=&]+)[\"']", 1, content);
+		for (String c : connects) {
+			switch(getLinkType(c)) {
+				case LINK_TYPE_IMAGE:
+					csp.getImageSource().addUrl(c);
+					break;
+				case LINK_TYPE_STYLE:
+					csp.getStyleSource().addUrl(c);
+					break;
+				case LINK_TYPE_SCRIPT:
+					csp.getScriptSource().addUrl(c);
+					break;
+				case LINK_TYPE_REQUEST:
+					csp.getConnectSource().addUrl(c);
+					break;
+				default:
+					break;
+			}
+		}
+	}
+	
+	private void handleImages(CSP csp, HttpMessage msg, Source source) {
+		// Images tags
+		List<Element> images = source.getAllElements("img");
+		for (Element image : images) {
+			String src = image.getAttributeValue("src");
+			if (src != null) {
+				csp.getImageSource().addUrl(src);
+			}
+		}
+	}
+	
+	private void handleMedia(CSP csp, HttpMessage msg, Source source) {
+		// Audio tags
+		List<Element> audio = source.getAllElements("audio");
```

Comment:
It can be extracted a method to handle the media tags, exactly the same code applies to audio/video.